src/main/java/net/neoforged/neoforge/resource/ResourcePackLoader.java

```diff
@@ -200,29 +203,26 @@ public static Pack.ResourcesSupplier createPackForMod(IModFileInfo mf) {
         return new PathPackResources.PathResourcesSupplier(mf.getFile().getSecureJar().getRootPath(), true);
     }
 
-    public static List<String> getDataPackNames() {
-        List<String> ids = new ArrayList<>(ModList.get().getModFiles().stream().filter(IModFileInfo::showAsDataPack)
+    public static List<String> getPackNames(PackType packType) {
+        List<String> ids = new ArrayList<>();
+        ids.add(packType == PackType.CLIENT_RESOURCES ? MOD_RESOURCES_ID : MOD_DATA_ID);
+        ids.addAll(ModList.get().getModFiles().stream().filter(packType == PackType.CLIENT_RESOURCES ? IModFileInfo::showAsResourcePack : IModFileInfo::showAsDataPack)
+                .filter(mf -> mf.requiredLanguageLoaders().stream().noneMatch(ls -> ls.languageName().equals("minecraft")))
                 .map(IModFileInfo::getFile)
-                .map(mf -> "mod:" + mf.getModInfos().get(0).getModId()).filter(n -> !n.equals("mod:minecraft"))
+                .map(mf -> "mod/" + mf.getModInfos().get(0).getModId())
                 .toList());
-        ids.add(MOD_DATA_ID);
         return ids;
     }
 
+    @Deprecated(forRemoval = true, since = "1.20.4")
+    public static List<String> getDataPackNames() {
+        return getPackNames(PackType.SERVER_DATA);
+    }
+
     public static <V> Comparator<Map.Entry<String, V>> getSorter(PackType packType) {
         List<String> order = new ArrayList<>();
         order.add("vanilla");
-        if (packType == PackType.CLIENT_RESOURCES) {
-            order.add(MOD_RESOURCES_ID);
-        } else {
-            order.add(MOD_DATA_ID);
-        }
-
-        ModList.get().getModFiles().stream()
-                .filter(mf -> mf.requiredLanguageLoaders().stream().noneMatch(ls -> ls.languageName().equals("minecraft")))
-                .map(e -> e.getMods().get(0).getModId())
-                .map(e -> "mod:" + e)
-                .forEach(order::add);
+        order.addAll(getPackNames(packType));
```

Comment:
Out of curiosity, couldn't this be changed to use an `ObjectLinkedOpenHashSet` instead of a list and a hashmap? To get the "index" in that case we would use `headSet(e.key()).size()`, and then a `contains` check rather than `getOrDefault`.

Reply:
Potentially; honestly, this sorter method should probably be deprecated anyways though. It's not used by neoforge, and only sorts the pack names of specifically mod-added, ungrouped packs.

Reply:
Yeah, just deprecated this; since this PR is going to be considered breaking I'll actually just remove all the old deprecated ResourcePackLoader methods once I can target 1.20.5